From 29f719565c0a73619692f96e6ffdaac4534783d9 Mon Sep 17 00:00:00 2001
From: Anil Kumar <Anil.Kumar@Wolfsonmicro.com>
Date: Tue, 5 Aug 2014 18:44:57 +0900
Subject: [PATCH 285/351] switch: arizona: Use pdata for HPDET short level
 impedance

Use the pdata/DT configuration of the short-circuit threshold
instead of a hardcoded value.

Change-Id: If8e53d8a7df42a418c9de0de22d344e0c6e67d14
Signed-off-by: Anil Kumar <Anil.Kumar@Wolfsonmicro.com>
Signed-off-by: Richard Fitzgerald <rf@opensource.wolfsonmicro.com>
---
 drivers/switch/switch-arizona.c |   18 ++++++++++++++----
 1 file changed, 14 insertions(+), 4 deletions(-)

diff --git a/drivers/switch/switch-arizona.c b/drivers/switch/switch-arizona.c
index e661445..64f00ec 100644
--- a/drivers/switch/switch-arizona.c
+++ b/drivers/switch/switch-arizona.c
@@ -64,6 +64,8 @@
 #define HP_NORMAL_IMPEDANCE     0
 #define HP_LOW_IMPEDANCE        1
 
+#define HP_LOW_IMPEDANCE_LIMIT 13
+
 struct arizona_extcon_info {
 	struct device *dev;
 	struct arizona *arizona;
@@ -286,7 +288,7 @@ static void arizona_extcon_do_magic(struct arizona_extcon_info *info,
 	}
 
 	/* Restore the desired state while not doing the magic */
-	if (!magic && arizona->hp_impedance > ARIZONA_HP_SHORT_IMPEDANCE) {
+	if (!magic && (arizona->hp_impedance > arizona->pdata.hpdet_short_circuit_imp)) {
 		ret = regmap_update_bits(arizona->regmap,
 					 ARIZONA_OUTPUT_ENABLES_1,
 					 ARIZONA_OUT1L_ENA |
@@ -695,11 +697,11 @@ int arizona_wm5110_tune_headphone(struct arizona_extcon_info *info,
 	const struct reg_default *patch;
 	int i, ret, size;
 
-	if (reading <= ARIZONA_HP_SHORT_IMPEDANCE) {
+	if (reading <= arizona->pdata.hpdet_short_circuit_imp) {
 		/* Headphones are always off here so just mark them */
 		dev_warn(arizona->dev, "Possible HP short, disabling\n");
 		return 0;
-	} else if (reading <= 13) {
+	} else if (reading <= HP_LOW_IMPEDANCE_LIMIT) {
 		if (info->hp_imp_level == HP_LOW_IMPEDANCE)
 			return 0;
 
@@ -745,7 +747,7 @@ int arizona_wm1814_tune_headphone(struct arizona_extcon_info *info,
 	const struct reg_default *patch;
 	int i, ret, size;
 
-	if (reading <= ARIZONA_HP_SHORT_IMPEDANCE) {
+	if (reading <= arizona->pdata.hpdet_short_circuit_imp) {
 		/* Headphones are always off here so just mark them */
 		dev_warn(arizona->dev, "Possible HP short, disabling\n");
 		return 0;
@@ -1750,6 +1752,9 @@ static int arizona_extcon_get_pdata(struct arizona *arizona)
 	arizona_of_read_u32(arizona, "wlf,hpdet-moisture-imp", false,
 			    &pdata->hpdet_moisture_imp);
 
+	arizona_of_read_u32(arizona, "wlf,hpdet-short-circuit-imp", false,
+			    &pdata->hpdet_short_circuit_imp);
+
 	arizona_of_read_u32(arizona, "wlf,hpdet-channel", false,
 			    &pdata->hpdet_channel);
 
@@ -1790,6 +1795,11 @@ static int arizona_extcon_probe(struct platform_device *pdev)
 
 	arizona_extcon_get_pdata(arizona);
 
+	if (pdata->hpdet_short_circuit_imp < 1)
+		pdata->hpdet_short_circuit_imp = ARIZONA_HP_SHORT_IMPEDANCE;
+	else if	(pdata->hpdet_short_circuit_imp >= HP_LOW_IMPEDANCE_LIMIT)
+		pdata->hpdet_short_circuit_imp = HP_LOW_IMPEDANCE_LIMIT - 1;
+
 	info = devm_kzalloc(&pdev->dev, sizeof(*info), GFP_KERNEL);
 	if (!info) {
 		dev_err(&pdev->dev, "Failed to allocate memory\n");
-- 
1.7.9.5

